patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
+ namespace videocapturemodule {
+ 
++HRESULT GetMediaCaptureForDevice(HSTRING device_id,
++                                  IMediaCapture** ppMediaCapture) {
++  HRESULT hr;
++
++  ComPtr<ICoreDispatcher> main_view_dispatcher;
++  ComPtr<IAsyncAction> async_action_ui_dispatcher;
++  ComPtr<IAsyncAction> async_action_media_capture;
++  ComPtr<IMediaCapture> media_capture;
++  AgileRef media_capture_agile;
++  AgileRef async_action_media_capture_agile;
++  Event event_wait_for_media_capture_async_action;
++  Event event_wait_for_media_capture_async_action_completion;
++  boolean has_thread_access;
++
++  // Acquires the main view dispacther (UI thread).
++  THR(GetMainViewDispatcher(&main_view_dispatcher));
++
++  // We'll be dispatching and waiting for code from the UI thread.
++  // Let's make sure this is not the UI thread.
++  THR(main_view_dispatcher->get_HasThreadAccess(&has_thread_access));
++  THR(has_thread_access ? E_FAIL : S_OK);
++
++  // The media capture device has to be initialized in the UI thread.
++  // The following event is used for letting this thread know that
++  // async_action_media_capture has been created.
++  event_wait_for_media_capture_async_action =
++      Event(::CreateEventEx(nullptr, nullptr, CREATE_EVENT_MANUAL_RESET,
++                            WRITE_OWNER | EVENT_ALL_ACCESS));
++  THR(event_wait_for_media_capture_async_action.IsValid()
++          ? S_OK
++          : HRESULT_FROM_WIN32(GetLastError()));
++
++  THR(ActivateInstance(
++      HStringReference(RuntimeClass_Windows_Media_Capture_MediaCapture).Get(),
++      media_capture.ReleaseAndGetAddressOf()));
++
++  THR(media_capture.AsAgile(&media_capture_agile));
++
++  // Dispatches the media capture initialization to the main view
++  // dispatcher.
++  THR(main_view_dispatcher->RunAsync(
++      CoreDispatcherPriority::CoreDispatcherPriority_Normal,
++      Callback<Implements<RuntimeClassFlags<Delegate>, IDispatchedHandler,
++                          FtmBase>>([&event_wait_for_media_capture_async_action,
++                                     &async_action_media_capture_agile,
++                                     &media_capture_agile,
++                                     &device_id]() -> HRESULT {
++        HRESULT hr;
++
++        ComPtr<IMediaCapture> media_capture;
++        ComPtr<IAsyncAction> media_capture_async;
++        ComPtr<IMediaCaptureInitializationSettings> init_settings;
++        ComPtr<IMediaCaptureInitializationSettings5> init_settings5;
++
++        THR(media_capture_agile.As(&media_capture));
++
++        // Creates the settings used to select which capture device will be
++        // used.
++        THR(ActivateInstance(
++            HStringReference(
++                RuntimeClass_Windows_Media_Capture_MediaCaptureInitializationSettings)
++                .Get(),
++            &init_settings));
++        THR(init_settings->put_StreamingCaptureMode(
++            StreamingCaptureMode::StreamingCaptureMode_AudioAndVideo));
```

Comment:
Currently on WebRTC UWP SDK the audio permission is required at startup because the audio device module (ADM; different from the capturer) initializes with the library (see `WebRtcVoiceEngine` init). The video permission on the other hand is only required if actually using video capture. This is the same on Android by the way; this is not specific to UWP, but rather to auth-based platforms in general. I don't know exactly how that works with the WinRT audio capturer you wrote but I am under the impression this won't change since this is an ADM limitation; so this audio permission is useless anyway, since it is requested much later, and you cannot reach that point in the code without crashing, at least with ADM1 (default). Feel free to leave for now, but I am pretty sure with or without won't make a difference, and if the current audio capture doesn't ask for permission then it is already susceptible to crashing the app, whether or not this change happens.

> Since most of the current use cases require audio and video

That is not always the case for Mixed Reality; several of our partners have apps with only one-way video, so video capture is not used on one peer. Some may have audio-only apps too. Although granted I do not know of any application using only video without audio.